app/scripts/controllers/permissions/index.js

```diff
@@ -188,33 +213,52 @@ export class PermissionsController {
    */
   async legacyExposeAccounts (origin, accounts) {
 
-    const permissions = {
-      eth_accounts: {},
+    // accounts are validated by finalizePermissionsRequest
+    if (typeof origin !== 'string' || !origin.length) {
+      throw new Error('Must provide non-empty string origin.')
+    }
+
+    const existingAccounts = await this.getAccounts(origin)
+
+    if (existingAccounts.length > 0) {
+      throw new Error(
+        'May not call legacyExposeAccounts on origin with exposed accounts.'
+      )
     }
 
-    await this.finalizePermissionsRequest(permissions, accounts)
+    const permissions = await this.finalizePermissionsRequest(
+      { eth_accounts: {} }, accounts
+    )
 
-    let error
     try {
+
       await new Promise((resolve, reject) => {
-        this.permissions.grantNewPermissions(origin, permissions, {}, (err) => (err ? resolve() : reject(err)))
+        this.permissions.grantNewPermissions(
```

Comment:
It's unfortunate that `grantNewPermissions` uses a callback rather than returning a Promise. This has behaves pretty unexpectedly now if an Error is thrown during that function - it'd just hang here indefinitely.

This is an `rpc-cap` issue though, not the fault of this PR.

Reply:
Ah, well, the reason for the callback signature was for `json-rpc-engine`-compatibility, and the use of that function is always Promise-wrapped, so such an error will always be caught.

The reason its usage is weird here is because `grantNewPermissions` wasn't intended for external consumption, and I've yet to write a wrapper in `rpc-cap` itself.

Reply:
>the use of that function is always Promise-wrapped, so such an error will always be caught.

It's not though - if any exception is thrown in `rpc-cap` in this function, it will hang. This Promise wrapper expects the callback to be called, and it won't be in that case.

You could guard against this in `rpc-cap` with a wrapper `try... catch` block around the whole function, but generally this problem is best avoided by using `async/await`, which handles errors the way you'd expect.

I'm not entirely sure what `json-rpc-engine` compatibility means in this case - do you mean that it was written to be functional as middleware?

Reply:
Resolved by means of [this fiddle](https://jsfiddle.net/3uknfdse/2/), which produces the following results in the console:
```FAILURE Error: unexpected error
    at cbFunc ((index):34)
    at (index):41
    at new Promise (<anonymous>)
    at wrappedCbFunc ((index):40)
    at main ((index):57)
    at (index):67
    at window.onload ((index):69)
SUCCESS 10
FAILURE too big
I am out of the loop
I am out of main```

Reply:
Right - I understand the middleware function signature, but I didn't follow why _this_ function in particular has that signature when it isn't used as middleware.

Was it _planned_ to be used as middleware but then never was? Or am I just missing where it's used 🤔

Reply:
Oh, it is used internally by `rpc-cap` in its middleware!

We didn't anticipate the need to use it externally (and indeed, it shouldn't be, hence the `@deprecated` tag)